Build failure: lldb_14

## Steps to reproduce

Steps to reproduce the behavior:

1. `nix shell github:nixos/nixpkgs/nixos-unstable#lldb_14`

## Build log

<!-- insert build log in code block in collapsable section -->

<details>

<summary>Build Log</summary>

```
error: builder for '/nix/store/7plaf8ff24b0a5xlnm2cxri94b23f4rm-lldb-14.0.6.drv' failed with exit code 1;
       last 25 log lines:
       > In file included from /nix/store/0nhi47d5ip48wprxnava6vv973zzzndf-libcxx-19.1.5-dev/include/c++/v1/__format/formatter_integral.h:21:
       > In file included from /nix/store/0nhi47d5ip48wprxnava6vv973zzzndf-libcxx-19.1.5-dev/include/c++/v1/__format/formatter_output.h:22:
       > In file included from /nix/store/0nhi47d5ip48wprxnava6vv973zzzndf-libcxx-19.1.5-dev/include/c++/v1/__format/parser_std_format_spec.h:39:
       > /nix/store/0nhi47d5ip48wprxnava6vv973zzzndf-libcxx-19.1.5-dev/include/c++/v1/string:2392:3: error: implicit instantiation of undefined template 'std::char_traits<unsigned 
char>'
       >  2392 |   traits_type::assign(__p[__n], value_type());
       >       |   ^
       > /nix/store/0nhi47d5ip48wprxnava6vv973zzzndf-libcxx-19.1.5-dev/include/c++/v1/__fwd/string.h:23:29: note: template is declared here
       >    23 | struct _LIBCPP_TEMPLATE_VIS char_traits;
       >       |                             ^
       > 4 errors generated.
       > [662/965] Building CXX object source/Plugins/Process/gdb-remote/CMakeFiles/lldbPluginProcessGDBRemote.dir/GDBRemoteRegisterContext.cpp.o
       > [663/965] Building CXX object source/Plugins/Process/gdb-remote/CMakeFiles/lldbPluginProcessGDBRemote.dir/GDBRemoteCommunicationServerPlatform.cpp.o
       > [664/965] Building LLDB Lua wrapper
       > [665/965] Building CXX object source/Plugins/Process/gdb-remote/CMakeFiles/lldbPluginProcessGDBRemote.dir/ProcessGDBRemoteLog.cpp.o
       > [666/965] Building CXX object source/Initialization/CMakeFiles/lldbInitialization.dir/SystemInitializer.cpp.o
       > [667/965] Building CXX object source/Initialization/CMakeFiles/lldbInitialization.dir/SystemInitializerCommon.cpp.o
       > [668/965] Building CXX object source/Plugins/Process/gdb-remote/CMakeFiles/lldbPluginProcessGDBRemote.dir/GDBRemoteCommunicationClient.cpp.o
       > [669/965] Building CXX object source/Plugins/Process/gdb-remote/CMakeFiles/lldbPluginProcessGDBRemote.dir/ThreadGDBRemote.cpp.o
       > [670/965] Building CXX object source/Initialization/CMakeFiles/lldbInitialization.dir/SystemLifetimeManager.cpp.o
       > [671/965] Building CXX object source/Plugins/TraceExporter/common/CMakeFiles/lldbPluginTraceExporterCommon.dir/TraceHTR.cpp.o
       > [672/965] Building CXX object source/Plugins/ABI/AArch64/CMakeFiles/lldbPluginABIAArch64.dir/ABIAArch64.cpp.o
       > [673/965] Building LLDB Python wrapper
       > Deprecated command line option: -py3. Ignored, this option is no longer supported.
       > [674/965] Building CXX object source/Plugins/Process/gdb-remote/CMakeFiles/lldbPluginProcessGDBRemote.dir/ProcessGDBRemote.cpp.o
       > ninja: build stopped: subcommand failed.
       For full logs, run 'nix log /nix/store/7plaf8ff24b0a5xlnm2cxri94b23f4rm-lldb-14.0.6.drv'.

```

</details>

## Additional context

<!-- Add any other context about the problem here. -->

This is on a `aarch64-darwin` (Apple Silicon M4 Pro).

## Metadata

<!-- Please insert the output of running `nix-shell -p nix-info --run "nix-info -m"` below this line -->

```
 - system: `"aarch64-darwin"`
 - host os: `Darwin 24.1.0, macOS 15.1`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.25.3`
 - channels(root): `"nixpkgs"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixpkgs`

```
## Notify maintainers

<!--
Please @ people who are in the `meta.maintainers` list of the offending package or module.
If in doubt, check `git blame` for whoever last touched something.
-->

---

Note for maintainers: Please tag this issue in your PR.

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/development/compilers/llvm/common/default.nix
+++ b/pkgs/development/compilers/llvm/common/default.nix
@@ -707,7 +707,15 @@ let
               stripLen = 1;
               hash = "sha256-QCGhsL/mi7610ZNb5SqxjRGjwJeK2rwtsFVGeG3PUGc=";
             })
-            ++ lib.optional (lib.versionOlder metadata.release_version "17") resourceDirPatch
+            ++ lib.optionals (lib.versionOlder metadata.release_version "17") [
+              resourceDirPatch
+              (fetchpatch {
+                name = "add-cstdio.patch";
+                url = "https://github.com/llvm/llvm-project/commit/73e15b5edb4fa4a77e68c299a6e3b21e610d351f.patch";
+                stripLen = 1;
+                hash = "sha256-eFcvxZaAuBsY/bda1h9212QevrXyvCHw8Cr9ngetDr0=";
+              })
+            ]
             ++ lib.optional (lib.versionOlder metadata.release_version "14") (
               metadata.getVersionFile "lldb/gnu-install-dirs.patch"
             )

--- a/pkgs/development/compilers/llvm/common/default.nix
+++ b/pkgs/development/compilers/llvm/common/default.nix
@@ -701,6 +701,12 @@ let
               # FIXME: do we need this after 15?
               (metadata.getVersionFile "lldb/procfs.patch")
             ]
+            ++ lib.optional (lib.versionOlder metadata.release_version "18") (fetchpatch {
+              name = "libcxx-19-char_traits.patch";
+              url = "https://github.com/llvm/llvm-project/commit/68744ffbdd7daac41da274eef9ac0d191e11c16d.patch";
+              stripLen = 1;
+              hash = "sha256-QCGhsL/mi7610ZNb5SqxjRGjwJeK2rwtsFVGeG3PUGc=";
+            })
             ++ lib.optional (lib.versionOlder metadata.release_version "17") resourceDirPatch
             ++ lib.optional (lib.versionOlder metadata.release_version "14") (
               metadata.getVersionFile "lldb/gnu-install-dirs.patch"

--- a/pkgs/development/compilers/llvm/common/lldb.nix
+++ b/pkgs/development/compilers/llvm/common/lldb.nix
@@ -127,13 +127,13 @@ stdenv.mkDerivation (rec {
 
   # TODO: cleanup with mass-rebuild
   installCheckPhase = ''
-    if [ ! -e $lib/${python3.sitePackages}/lldb/_lldb*.so ] ; then
+    if [ ! -e ''${!outputLib}/${python3.sitePackages}/lldb/_lldb*.so ] ; then
         echo "ERROR: python files not installed where expected!";
         return 1;
     fi
   '' # Something lua is built on older versions but this file doesn't exist.
   + lib.optionalString (lib.versionAtLeast release_version "14") ''
-    if [ ! -e "$lib/lib/lua/${lua5_3.luaversion}/lldb.so" ] ; then
+    if [ ! -e "''${!outputLib}/lib/lua/${lua5_3.luaversion}/lldb.so" ] ; then
         echo "ERROR: lua files not installed where expected!";
         return 1;
     fi

